Draft for review: steps for building and deploying a smart contract (#200)
input-output-hk/marlowe-doc: docs/getting-started/steps-to-build-deploy-contract.md

---

> title: Steps to build and deploy a smart contract

Reviewer: Consider changing to gerund for consistency with the writing guidelines:
'Building and deploying a smart contract'

Author: Good point. I'll rename the file as well to "building-deploying-smart-contract.md".

sidebar_position: 8
---

> This is an explanation of the process of building a smart contract in Marlowe from start to deployment on the preview or pre-production network of the Cardano blockchain. 

Reviewer: ```suggestion
This is an explanation of the process of building a smart contract in Marlowe from start to deployment on Cardano's preview or pre-production testing networks. 
```

> 

Reviewer: Would be great to add prerequisites. Please talk to SMEs to define and add those.

Reviewer: Edit: since the page breaks down content into explainers and guidelines, it would be better to add prerequisites below in the development part.

Author: Good point about prerequisites. I will ask for more input on this topic. I recommend having a single section for prerequisites near the top of the page that covers all the essential aspects of caution for both contract building and deployment. I'll start with this: 

## Prerequisites

* See [**Learning Marlowe**](https://docs.marlowe.iohk.io/docs/getting-started/learning-marlowe) to familiarize yourself with the tutorials and other recommended topics. 
* See the tutorial about [**checking the safety of a Marlowe contract**](https://github.com/input-output-hk/marlowe-starter-kit/blob/main/lessons/07-safety/07-safety.ipynb)
* See also [**Using Marlowe safely**](https://github.com/input-output-hk/marlowe-starter-kit/blob/main/docs/using-marlowe-safely.md)
* [**Considerations before you deploy**](https://docs.marlowe.iohk.io/docs/getting-started/runner#considerations-before-you-deploy)

## Building a smart contract

Creating a smart contract involves the following stages: 
* Conceptualization
* Design
* Development

> * Simulation and testing

Reviewer: ```suggestion
* Simulation and testing.
```

### Conceptualization

Take the time to think through what exactly you want your smart contract to accomplish. Carefully consider what the key advantages would be of conducting your transaction with a smart contract on the blockchain. Here are some essential questions to ask yourself as you conceptualize your contract: 
* What is the central purpose of your smart contract? For example, it could be to facilitate a transaction or to enforce an agreement. 
* Who are your participants? 
* What roles and responsibilities will each of them they have? 
* What conditions need to be met along the way in order for the contract to proceed through each stage? 
* Think about the logic that the contract will need to take into account in the scenario where one or more participants do not fulfill their obligations. How will you protect your interests? 
* What conditions need to be part of the contract so that each participant is assured that the terms are fair and reasonable? 
* Will you need to take into account any external data or events? If so, how will you handle those factors? 

### Design

After completing the conceptualization stage, you can start designing your smart contract. 

> - **Structure**: Layout the structure of your contract. Think about the sequence of actions that will be involved. How will funds need to be handled at each stage? 

Reviewer: Please change all the colons to full stops for consistency with the writing guidelines.

> - **Roles**: Define each participant’s role, such as buyer, seller, and intermediary. What actions will each role need to take for each stage of the contract? 

Reviewer: ```suggestion
- **Roles**: Define each participant’s role, such as _buyer_, _seller_, and _intermediary_. What actions will each role need to take for each stage of the contract? 
```

Please see these resources for more detailed information about working with the Playground: 

* [**Developer tools > Marlowe Playground**](https://docs.marlowe.iohk.io/docs/developer-tools/playground)
* [**Tutorials > Overview of the Marlowe Playground**](https://docs.marlowe.iohk.io/tutorials/concepts/playground-overview)

### Simulation and testing

Devote time and effort towards testing your contract in the Playground. Simulation and testing are a critical part of your smart contract development. Simulating your contract allows you to test various possible pathways and scenarios to make sure that your contract behaves as expected. 

The simulation will help to reveal any errors in your contract such as failing to make a payment. You can also verify that a refund occurs if certain conditions are met. If you do not include a timeout, simulating the contract will help to reveal this. For example, if a participant fails to make a deposit, the contract should include a timeout. 

> #### To simulate your contract

Reviewer: ```suggestion
#### To simulate your contract:
```

1. With your contract open in the Playground, select ‘Send to Simulator’ in the upper right area. A panel opens where you can specify deadline parameters and ada amounts. 
2. Step through the contract’s execution and test various choices, examining different potential pathways. This will give you insight into how your contract will behave under different conditions. 

## Deploying a smart contract

> Use Marlowe Runner to deploy your smart contract to the blockchain. Marlowe Runner streamlines and simiplifies the process of deploying smart contracts on Cardano. It eliminates the need for low-level tools and deploying your own backend Runtime instance. 

Reviewer: Please add a link to Marlowe Runner

Author: I'll add these two links below the paragraph: 

* [**Access Runner on the preview network**](https://preview.runner.marlowe.iohk.io/)
* [**Access Runner on the pre-production network**](https://preprod.runner.marlowe.iohk.io/)

> #### To deploy your contract

Reviewer: ```suggestion
#### To deploy your contract:
```

1. With your contract open in the Playground, select ‘Send to Simulator,’ then ‘Export to Marlowe Runner.’ A window displays prompting you to select either the pre-production or preview network for deployment. After you make your selection, Playground generates and exports the JSON file directly to Marlowe Runner. Runner automatically displays in a browser window and prompts you to choose a wallet. 
2. Make your wallet selection. Make sure that your wallet is connected to the same network that you selected for your contract deployment. Runner displays the source code for your contract. 
3. In the ‘Add tags’ field, enter a name for your contract. 
4. Click ‘Submit contract.’ Depending on the details of your contract, you may be prompted to specify addresses for borrower and lender, or you may be prompted to confirm the transaction with your wallet. 

> 5. Follow the prompts. The Runner displays a list view that shows the current status of your contract. The status shown under the Actions column will update as your contract progresses. 

Reviewer: ```suggestion
5. Follow the prompts. Runner displays a list view that shows the current status of your contract. The status shown under the 'Actions' column will update as your contract progresses. 
```

> 6. When ‘Advance’ displays under the Action column for your contract, select ‘Advance’ to proceed. The Source graph page displays. The source graph will show the possible execution paths of your contract. 

Reviewer: ```suggestion
6. When ‘Advance’ displays under the 'Actions' column for your contract, select ‘Advance’ to proceed. The source graph page displays. The source graph will show the possible execution paths of your contract. 
```

7. Below the source graph, the next available action will be reflected, such as a deposit, for example. Runner is designed to prompt you through the execution of your contract. 
8. Continue to follow the prompts, using your wallet password to authorize transactions for your role. 

> From the Runner page that lists pending transactions, when the 'Action' column indicates a status of ‘Complete,’ select the Contract Id link to view your transaction in the source graph page. It will show the execution path of your contract. 

Reviewer: Action or Actions column? I noticed the use of both, please double-check and update. Thank you.

Author: Good catch. I checked, and it is the 'Actions' column.

> From the Runner page that lists pending transactions, when the 'Action' column indicates a status of ‘Complete,’ select the Contract Id link to view your transaction in the source graph page. It will show the execution path of your contract. 

Reviewer: ```suggestion
From the Runner page that lists pending transactions, when the 'Action' column indicates a status of ‘Complete,’ select the contract Id link to view your transaction in the source graph page. It will show the execution path of your contract. 
```
